### DISCUSSION TITLE
Can't Deploy To Vercel Due To Build Error

### DISCUSSION THREAD

**Discussion by darrylmorley** [vercel/next.js#21838] [Help] [answered]
Hi,

I keep getting build errors on a site that I'm building, the error is:

```

01:41:05.672 | info  - Compiled successfully
-- | --
01:41:05.672 | info  - Collecting page data...
01:41:06.983 | undefined
01:41:06.985 | > Build error occurred
01:41:06.987 | TypeError: Cannot read property 'data' of undefined
01:41:06.987 | at getStaticPaths (/vercel/workpath0/.next/serverless/pages/ammo/[id].js:1239:31)
01:41:06.987 | at processTicksAndRejections (internal/process/task_queues.js:97:5)
01:41:06.987 | at async buildStaticPaths (/vercel/workpath0/node_modules/next/dist/build/utils.js:17:80)
01:41:06.987 | at async Object.isPageStatic (/vercel/workpath0/node_modules/next/dist/build/utils.js:24:549) {
01:41:06.987 | type: 'TypeError'
01:41:06.987 | }
01:41:07.038 | error Command failed with exit code 1.
01:41:07.038 | info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
01:41:07.051 | Error: Command "yarn run build" exited with 1
01:41:11.637 | Done with "package.json"
```

This is my getStaticPaths() from that file:
```
export async function getStaticPaths() {
  const data = await getAmmo()
  const pathData = await data.data.Item
  
  const paths = pathData.map(item => ({
    params: { id: item.itemID }
  }))

  return { paths, fallback: true }
}
```
I can't see anything wrong here and seeing as I'm not receiving an error from getAmmo() I have no idea how to try and resolve this.

Can anybody please offer any advice?

Thanks

My getAmmo() function just in case:

```
export async function getAmmo() {
  let axiosConfig = await getHeader()
  return await axios.get(`Item.json?load_relations=["ItemShops", "Category", "Images", "ItemECommerce"]&ItemShops.qoh=%3E%3D,1&Category.categoryID=IN,[23,46,218]&ItemECommerce.listOnStore=true`, axiosConfig).catch(err => console.error(err.data))
}
```

Everything works fine with 'yarn dev'

Edit: console.log(paths) in dev:
```
[
  { params: { id: '31' } },
  { params: { id: '57' } },
  { params: { id: '71' } },
  { params: { id: '72' } },
  { params: { id: '972' } },
  { params: { id: '989' } },
  { params: { id: '1240' } },
  { params: { id: '1713' } },
  { params: { id: '1844' } },
  { params: { id: '1849' } },
  { params: { id: '1927' } },
  { params: { id: '2362' } },
  { params: { id: '2363' } },
  { params: { id: '2377' } },
  { params: { id: '2397' } },
  { params: { id: '2432' } },
  { params: { id: '2458' } },
  { params: { id: '2461' } },
  { params: { id: '2464' } },
  { params: { id: '2465' } },
  { params: { id: '2467' } },
  { params: { id: '2469' } },
  { params: { id: '2470' } },
  { params: { id: '2682' } },
  { params: { id: '2877' } },
  { params: { id: '2883' } },
  { params: { id: '3250' } },
  { params: { id: '3252' } },
  { params: { id: '3272' } },
  { params: { id: '3290' } },
  { params: { id: '3923' } },
  { params: { id: '4609' } },
  { params: { id: '4611' } },
  { params: { id: '5223' } },
  { params: { id: '6902' } },
  { params: { id: '6903' } },
  { params: { id: '7071' } },
  { params: { id: '7072' } }
]
```

**Comment by enochndika** [comment]
I also encountered the same error with getStaticProps, I solved it by first checking if data exist

```
  return (
    <>
      {data && (
        <div>
          your code
        </div>
      )}
    </>
  );
```

**Reply by darrylmorley** [reply]
Could be I guess, but the error is saying that it's not getting the data in getStaticPaths at build time.

**Reply by darrylmorley** [reply]
Hi @enochndika,

Thanks for the input. I did try your solution, unfortunately still the same error. Thanks for replying though 👍

**Comment by leerob** [comment]
1. Have you added any necessary environment variables to Vercel to connect to this service?
2. If you run `next build && next start` locally, does it also work?

**Reply by darrylmorley** [reply]
Hi @leerob,

Your Youtube and blogs have been really helpful, thanks!

My issue:

1. Yes all added.
2. No, it's not building locally either, the same error response.

Now, I had this same issue with a different site last night. That site built fine again this morning but is pulling much less data. I comapre because the two sites are pulling data from the same API.

It's as though the build process just charges ahead without waiting for the data from the API, is this the case? And, if so, is there away to force the build process to wait for the data?

**Reply by leerob** [reply]
Yeah sounds like an issue fetching your data then! Not sure what service you're trying to connect to, but maybe you're missing an `await` somewhere if it's not waiting?

**Reply by darrylmorley** [reply]
I made some changes as per your suggestions and managed to get a local build to work 👍. Unfortuantely all builds since have failed still 🤦‍♂️. I'm mulling switching to getInitialProps at this point, but not the ideal solution.

**Comment by michrome** [comment]
Hey @darrylmorley! Looking at this line:

```js
return await axios.get(`Item.json?load_relations=["ItemShops", "Category", "Images", "ItemECommerce"]&ItemShops.qoh=%3E%3D,1&Category.categoryID=IN,[23,46,218]&ItemECommerce.listOnStore=true`, axiosConfig).catch(err => console.error(err.data))
```

It will catch an error response but it _won't_ catch a non-error response that's missing information you require.

For example, I can imagine `{}` being a non-error response but it doesn't contain a property named `data`. However, this line makes the assumption all non-error responses will contain a `data` property:

```js
const pathData = await data.data.Item
```

My recommendation is to have `function getAmmo()` do _something_ if its return value doesn't contain a property named `data`. That _something_ could be throw an error or set a default value for `data`, for example.


Hope that helps?

**Reply by darrylmorley** [reply]
Hey @michrome, 

I'll make some changes and give it a shot. Thanks for the pointer!

**Comment by darrylmorley** [comment, marked as answer]
I think it was an issue with one of my env keys, I must have cut off a character or something when copying and pasting.
